src/SubstanceCore.hs

```diff
@@ -0,0 +1,255 @@
+type SubProg = [SubStmt]
+type SubObjDiv = ([SubDecl], [SubConstr])
+
+-- | Declaration of Substance objects
+data SubDecl = SubDeclConst Type Var
+               deriving (Show, Eq, Typeable)
+
+-- | Declaration of Substance constaints
+data SubConstr = SubConstrConst Predicate [Var]
+                 deriving (Show, Eq, Typeable)
+
+-- --------------------------------------- Substance Lexer -------------------------------------
+
+-- | TODO: Some of the lexing functions are alreadt implemented at Utils.hs, Merge it and avoid duplications!
+type Parser = Parsec Void String
+
+-- | 'lineComment' and 'blockComment' are the two styles of commenting in Penrose and in the DSLL
+lineComment, blockComment :: Parser ()
+lineComment  = L.skipLineComment "--"
+blockComment = L.skipBlockComment "/*" "*/"
+
+
+-- | A strict space consumer. 'sc' only eats space and tab characters. It does __not__ eat newlines.
+sc :: Parser ()
+sc = L.space (void $ takeWhile1P Nothing f) lineComment empty
+  where
+    f x = x == ' ' || x == '\t'
+
+-- | A normal space consumer. 'scn' consumes all whitespaces __including__ newlines.
+scn :: Parser ()
+scn = L.space space1 lineComment blockComment
+
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme sc
+
+-- A warper to make sure that we comsume whitespace after every lexeme
+symbol :: String -> Parser String
+symbol = L.symbol sc
+
+newline' :: Parser ()
+newline' = newline >> scn
+
+-- Pill out the left and right parens from the input and roll-out the parsing
+parens :: Parser a -> Parser a
+parens = between (symbol "(") (symbol ")")
+
+listOut :: Parser a -> Parser a
+listOut = between (symbol "[") (symbol "]")
+
+-- Define characters wich are part of the syntax
+colon, arrow, comma, rsbrac, lsbrac, lparen, rparen,eq :: Parser ()
+colon  = void (symbol ":")
+comma  = void (symbol ",")
+arrow  = void (symbol "->")
+lsbrac = void (symbol "[")
+rsbrac = void (symbol "]")
+lparen = void (symbol "(")
+rparen = void (symbol ")")
+eq     = void (symbol "=")
+
+-- Reserverd word handling
+rword :: String -> Parser ()
+rword w = (lexeme . try) (string w *> notFollowedBy alphaNumChar)
+
+ -- The list of reserved words in DSLL
+rws :: [String]
+rws = [] -- TODO: Doesn't seems like there is any reserved word here, should seperate reserved words in Utils.hs for each PL
+
+-- Identifiers handling
+identifier :: Parser String
+identifier = (lexeme . try) (p >>= check)
+  where
+    p       = (:) <$> letterChar <*> many alphaNumChar
+    check x = if x `elem` rws
+                then fail $ "keyword " ++ show x ++ " cannot be an identifier"
+                else return x
+
+--------------------------------------- DSLL Parser -------------------------------------
+
+-- | 'substanceParser' is the top-level parser function. The parser contains a list of functions
+--    that parse small parts of the language. When parsing a source program, these functions are invoked in a top-down manner.
+substanceParser :: Parser [SubStmt]
```

Comment:
maybe clarify that the Substance parser does not depend on the DSLL info, just the Substance typechecker does